cpp/src/arrow/compute/kernels/scalar_if_else_test.cc

```diff
@@ -624,6 +624,250 @@ TYPED_TEST(TestCaseWhenNumeric, ListOfType) {
+  // input when outputting null)
+  CheckScalar("case_when", {MakeStruct({Datum(true), Datum(false)}), values1, values2},
+              values1);
+  CheckScalar("case_when", {MakeStruct({Datum(false), Datum(true)}), values1, values2},
+              values2);
+  CheckScalar("case_when", {MakeStruct({Datum(false), Datum(false)}), values1, values2},
+              values1_null);
+  CheckScalar("case_when", {MakeStruct({Datum(false), Datum(false)}), values2, values1},
+              values2_null);
+
+  // For array conditions, we always borrow the dictionary of the first input
+  CheckScalar("case_when", {MakeStruct({cond1, cond2}), values1, values2},
+              DictArrayFromJSON(type, "[0, null, null, null]", dict1));
+  CheckScalar("case_when", {MakeStruct({cond1, cond2}), values1, values2, values1},
+              DictArrayFromJSON(type, "[0, null, null, 1]", dict1));
+
+  // When mixing dictionaries, we try to map other dictionaries onto the first one
+  // Don't check the scalar cases since we don't remap dictionaries in that case
+  CheckScalarNonRecursive(
+      "case_when",
+      {MakeStruct({ArrayFromJSON(boolean(), "[true, true, false, false]")}), values1,
+       values2},
+      DictArrayFromJSON(type, "[0, null, null, 2]", dict1));
+  CheckScalarNonRecursive(
+      "case_when",
+      {MakeStruct({ArrayFromJSON(boolean(), "[true, true, false, false]"),
+                   ArrayFromJSON(boolean(), "[true, false, true, false]")}),
+       values1, values2},
+      DictArrayFromJSON(type, "[0, null, null, null]", dict1));
+  CheckScalarNonRecursive(
+      "case_when",
+      {MakeStruct({ArrayFromJSON(boolean(), "[false, false, false, false]"),
+                   ArrayFromJSON(boolean(), "[true, true, true, true]")}),
+       values1, values3},
+      DictArrayFromJSON(type, "[3, 0, 0, 2]", dict1));
+  CheckScalarNonRecursive(
+      "case_when",
+      {MakeStruct({ArrayFromJSON(boolean(), "[null, null, null, true]"),
+                   ArrayFromJSON(boolean(), "[true, true, true, true]")}),
+       values1, values3},
+      DictArrayFromJSON(type, "[3, 0, 0, 1]", dict1));
+  CheckScalarNonRecursive(
+      "case_when",
+      {
+          MakeStruct({ArrayFromJSON(boolean(), "[true, true, false, false]")}),
+          DictScalarFromJSON(type, "0", dict1),
+          DictScalarFromJSON(type, "0", dict2),
+      },
+      DictArrayFromJSON(type, "[0, 0, 2, 2]", dict1));
+  CheckScalarNonRecursive(
+      "case_when",
+      {
+          MakeStruct({ArrayFromJSON(boolean(), "[true, true, false, false]"),
+                      ArrayFromJSON(boolean(), "[false, false, true, true]")}),
+          DictScalarFromJSON(type, "0", dict1),
+          DictScalarFromJSON(type, "0", dict2),
+      },
+      DictArrayFromJSON(type, "[0, 0, 2, 2]", dict1));
+
+  // If we can't map values from a dictionary, then raise an error
+  // Unmappable value is in the else clause
```

Comment:
I'm curious: why don't we unify dictionaries instead? It would sound more useful to me. I don't see any reason for the first input to have a particular status, is there?

Reply:
I had mostly tried to emulate the R/dplyr behavior as closely as possible: https://github.com/apache/arrow/pull/10724#discussion_r682676388

But unification is honestly probably easier to implement for us, so I can switch to that instead.